README.md
# Prosessilouhinta

Process mining (Finnish prosessilouhinta) from eventlogs.

Given a CSV file containing an eventlog with case IDs, acitivities/tasks, user IDs, 
and timestamps some evaluations are performed that may assist in process mining.

[License: MIT](https://git.sr.ht/~sthagen/prosessilouhinta/tree/default/item/LICENSE)

Third party dependencies are documented in the folder [third-party](docs/third-party/README.md).

[![version](https://img.shields.io/pypi/v/prosessilouhinta.svg?style=flat)](https://pypi.python.org/pypi/prosessilouhinta/)
[![downloads](https://pepy.tech/badge/prosessilouhinta/month)](https://pepy.tech/project/prosessilouhinta)
[![wheel](https://img.shields.io/pypi/wheel/prosessilouhinta.svg?style=flat)](https://pypi.python.org/pypi/prosessilouhinta/)
[![supported-versions](https://img.shields.io/pypi/pyversions/prosessilouhinta.svg?style=flat)](https://pypi.python.org/pypi/prosessilouhinta/)
[![supported-implementations](https://img.shields.io/pypi/implementation/prosessilouhinta.svg?style=flat)](https://pypi.python.org/pypi/prosessilouhinta/)
[![maintenance-status](https://img.shields.io/github/commit-activity/y/sthagen/prosessilouhinta.svg?style=flat)](https://git.sr.ht/~sthagen/prosessilouhinta/log)

## Documentation

User and developer [documentation of prosessilouhinta](https://codes.dilettant.life/docs/prosessilouhinta).

## Bug Tracker

Any feature requests or bug reports shall go to the [todos of prosessilouhinta](https://todo.sr.ht/~sthagen/prosessilouhinta).

## Primary Source repository

The main source of `prosessilouhinta` is on a mountain in central Switzerland.
We use distributed version control (git).
There is no central hub.
Every clone can become a new source for the benefit of all.
The preferred public clones of `prosessilouhinta` are:

* [on codeberg](https://codeberg.org/sthagen/prosessilouhinta) - a democratic community-driven, non-profit software development platform operated by Codeberg e.V.
* [at sourcehut](https://git.sr.ht/~sthagen/prosessilouhinta) - a collection of tools useful for software development.

## Contributions

Please do not submit "pull requests" (I found no way to disable that "feature" on GitHub).
If you like to share small changes under the repositories license please kindly do so by sending a patchset.
You can either send such a patchset per email using [git send-email](https://git-send-email.io) or 
if you are a sourcehut user by selecting "Prepare a patchset" on the summary page of your fork at [sourcehut](https://git.sr.ht/).

## Status

Experimental.

**Note**: The default branch is `default`.


docs/changes.md
# Changes

2022.11.13
:    * Added Critical Path Analysis (CPA) implementation with tests
* Raised the bar to python versions greater or equal to 3.9

2022.09.24
:    * Added cleansing of site folder to clean target
* Added developer dep for docstring style check from cartesian strategy source
* Added editor artifacts to git ignores
* Added experimental pytest plugins to test deps from cartesian strategy source
* Added new dev and test deps
* Added packaging dep and bumped dev deps
* Added test coverage page/link to docs
* Bumped impl, dev, and test deps
* Dropped dev deps (prospector==1.7.7 as prospector still insists on older mccabe)
* End of line sanity across platforms (LF)
* Excluded tests from packaging
* Exempted the matrix variable names from the dump linter mindset
* Fixed #4 by adding namespace prefix to control_flow call
* Fixed typos in documentation # 3
* Fixed, updated, and harmonized make file
* Made README and docs landing page lobby for the distributed in dvcs (git), codeberg, and sourcehut
* Migrated from setup.cfg to pyproject.toml for packaging
* Patched README and docs landing page to provide the original codeberg blurb from https://docs.codeberg.org/getting-started/what-is-codeberg/
* Removed dev tool that triggers false positive on license deviations at libraries.io and adapted to new test folder name
* Removed final traces of github from the docs
* Updated editor config

## 2021

2021.11.27
:    * Initial release to pypi


test/__init__.py
# -*- coding: utf-8 -*-
# pylint: disable=missing-docstring

__version__ = '0'

"""Some packages' test package initialisation."""


test/test_cli.py
import pathlib

from typer.testing import CliRunner

import prosessilouhinta
import prosessilouhinta.cli as cli
from prosessilouhinta.cli import app

BASIC_FIXTURES_PATH = pathlib.Path('test', 'fixtures', 'basic')

runner = CliRunner()


def test_app_version():
    result = runner.invoke(app, ['version'])
    assert result.exit_code == 0
    assert cli.APP_NAME in result.stdout
    assert prosessilouhinta.__version__ in result.stdout


def test_app_extract():
    result = runner.invoke(app, ['extract'])
    assert result.exit_code == 0


def test_app_unknown():
    result = runner.invoke(app, ['unknown', 'STDIN', 'does-not-exist'])
    assert result.exit_code == 2


def test_cli_main():
    message = 'source is no file'
    result = runner.invoke(app, ['extract', 'no_file_there'])
    assert result.exit_code == 1
    assert message in result.stdout


def test_cli_main_unknown_command():
    message = 'No such command'
    result = runner.invoke(app, ['unknown', 'STDIN', 'STDOUT', 'DRYRUN'])
    assert result.exit_code == 2
    assert message in result.stdout


def test_cli_main_source_does_not_exist():
    message = 'source is no file'
    result = runner.invoke(app, ['extract', 'source-does-not-exist', 'STDOUT', '-n'])
    assert result.exit_code == 1
    assert message in result.stdout


def test_cli_main_target_does_exist():
    message = 'target file exists'
    existing_file = str(BASIC_FIXTURES_PATH / 'existing-out-file.whatever')
    result = runner.invoke(app, ['extract', existing_file, existing_file, '--dryrun'])
    assert result.exit_code == 1
    assert message in result.stdout


def test_cli_main_verifier_passes():
    message = r'dryrun requested'
    existing_file = str(BASIC_FIXTURES_PATH / 'existing-out-file.whatever')
    result = runner.invoke(app, ['extract', existing_file, '-o', 'target-does-not-exist', '-n'])
    assert result.exit_code == 0
    assert message in result.stdout


def test_cli_main_too_few_columns():
    message = 'c1,t1,2021-11-27 12:34:56'
    # error = 'not enough values to unpack (expected 4, got 3)'
    result = runner.invoke(app, ['extract', str(BASIC_FIXTURES_PATH / 'single-too-short-data-line.csv')])
    assert result.exit_code == 1
    assert message in result.stdout


test/test_cpa.py
import json
import operator
import pathlib

import prosessilouhinta.cpa as cpa

CPA_MICRO_FIXTURE_PATH = pathlib.Path('test', 'fixtures', 'basic', 'cpa-micro.json')
CPA_SMALL_FIXTURE_PATH = pathlib.Path('test', 'fixtures', 'basic', 'cpa-small.json')
CPA_SMALL_FIXTURE_CRITICAL_PATH = pathlib.Path('test', 'fixtures', 'basic', 'cpa-small-critical-path-nodes.json')

with open(CPA_SMALL_FIXTURE_CRITICAL_PATH, 'rt', encoding='utf-8') as handle:
    CPA_SMALL_CRTICAL_PATH_EXPECTED = json.load(handle)


def test_cpa_cyclic_empty_dgraph():
    empty = {}
    assert cpa.cyclic(empty) is False


def test_cpa_cyclic_on_minimal_cyclic_dgraph():
    good = {'a': ('a',)}
    assert cpa.cyclic(good) is True


def test_cpa_cyclic_on_acyclic_dgraph():
    good = {'a': ('z',), 'z': ('b',), 'b': ('y', 'c')}
    assert cpa.cyclic(good) is False


def test_cpa_cyclic_on_cyclic_dgraph():
    bad = {'a': ('z',), 'z': ('b',), 'b': ('y', 'z')}
    assert cpa.cyclic(bad) is True


def test_cpa_nodes():
    nodes = set()
    n1 = cpa.Node(name=1, duration=1)
    n2 = cpa.Node(name=2, duration=1)
    n2a = cpa.Node(name=2, duration=1)
    assert n2 == n2a
    nodes.add(n1)
    nodes.add(n2)
    nodes.add(n2a)
    assert len(nodes) == 2

    parent = cpa.Node('parent')
    assert len(parent.nodes) == 0
    assert n1 not in parent.nodes
    parent.add(n1)
    assert len(parent.nodes) == 1
    assert n1 in parent.nodes
    assert n2 not in parent.nodes
    parent.add(n2)
    assert len(parent.nodes) == 2
    assert n1 in parent.nodes
    assert n2 in parent.nodes


def test_cpa_nodes_in_acyclic_dgraph():
    p = cpa.Node('acyclic')
    a = p.add(cpa.Node('A', duration=3))
    b = p.add(cpa.Node('B', duration=3, lag=0))
    c = p.add(cpa.Node('C', duration=4, lag=0))
    d = p.add(cpa.Node('D', duration=6, lag=0))
    e = p.add(cpa.Node('E', duration=5, lag=0))
    p.link(a, b).link(a, c).link(a, d).link(b, e).link(c, e).link(d, e)
    assert p.is_acyclic() is True


def test_cpa_nodes_in_cyclic_dgraph():
    p = cpa.Node('cyclic')
    a = p.add(cpa.Node('A', duration=3))
    b = p.add(cpa.Node('B', duration=3, lag=0))
    c = p.add(cpa.Node('C', duration=4, lag=0))
    d = p.add(cpa.Node('D', duration=6, lag=0))
    e = p.add(cpa.Node('E', duration=5, lag=0))
    p.link(a, b).link(a, c).link(a, d).link(b, e).link(c, e).link(d, e).link(e, a)
    assert p.is_acyclic() is False


def test_cpa_of_micro_project():
    p = cpa.Node('micro')
    a = p.add(cpa.Node('A', duration=3))
    b = p.add(cpa.Node('B', duration=3, lag=0))
    c = p.add(cpa.Node('C', duration=4, lag=0))
    d = p.add(cpa.Node('D', duration=6, lag=0))
    e = p.add(cpa.Node('E', duration=5, lag=0))
    # a -> b -> e
    # a -> c -> e
    # a -> d -> e
    p.link(a, b).link(a, c).link(a, d).link(b, e).link(c, e).link(d, e).update_all()

    aons = {}
    for node in sorted(p.nodes, key=operator.attrgetter('name')):
        aons[node.name] = node.activity_on_node_text()

    cp = p.get_critical_path()
    assert p is not None
    aon_p = """\
+----------------+
|     DUR=14     |
+----------------+
|ES=0|     |EF=14|
|----|micro|-----|
|LS=0|     |LF=14|
+----------------+
|    DRAG=n/a    |
+----------------+
"""
    assert p.activity_on_node_text() == aon_p
    p_aon_map = {
        'duration': 14,
        'earliest_start': 0,
        'earliest_finish': 14,
        'name': 'micro',
        'latest_start': 0,
        'latest_finish': 14,
        'drag': None,
    }
    the_data = p.activity_on_node()
    assert the_data == p_aon_map
    assert json.dumps(the_data) == json.dumps(p_aon_map)  # Just to be sure :-)

    assert p.activity_on_node_strings() == tuple(row for row in aon_p.split('\n') if row)

    assert a is not None
    aon_a = """\
+-----------+
|   DUR=3   |
+-----------+
|ES=0| |EF=3|
|----|A|----|
|LS=0| |LF=3|
+-----------+
|  DRAG=n/a |
+-----------+
"""
    assert aons[a.name] == aon_a

    assert b is not None
    aon_b = """\
+-----------+
|   DUR=3   |
+-----------+
|ES=3| |EF=6|
|----|B|----|
|LS=6| |LF=9|
+-----------+
|  DRAG=n/a |
+-----------+
"""
    assert aons[b.name] == aon_b

    assert c is not None
    aon_c = """\
+-----------+
|   DUR=4   |
+-----------+
|ES=3| |EF=7|
|----|C|----|
|LS=5| |LF=9|
+-----------+
|  DRAG=n/a |
+-----------+
"""
    assert aons[c.name] == aon_c

    aon_d = """\
+-----------+
|   DUR=6   |
+-----------+
|ES=3| |EF=9|
|----|D|----|
|LS=3| |LF=9|
+-----------+
|  DRAG=n/a |
+-----------+
"""
    assert d is not None
    assert aons[d.name] == aon_d

    assert e is not None
    aon_e = """\
+------------+
|   DUR=5    |
+------------+
|ES=9| |EF=14|
|----|E|-----|
|LS=9| |LF=14|
+------------+
|  DRAG=n/a  |
+------------+
"""
    assert aons[e.name] == aon_e

    assert cp == [a, d, e]
    a_el = a.activity_on_node_strings_with_link()
    d_el = d.activity_on_node_strings_with_link()
    e_el = e.activity_on_node_strings()

    aon_dia = tuple(f'{fir}{sec}{thi}' for fir, sec, thi in zip(a_el, d_el, e_el))
    expected_diagram = """\
+-----------+    +-----------+    +------------+
|   DUR=3   |    |   DUR=6   |    |   DUR=5    |
+-----------+    +-----------+    +------------+
|ES=0| |EF=3|    |ES=3| |EF=9|    |ES=9| |EF=14|
|----|A|----| => |----|D|----| => |----|E|-----|
|LS=0| |LF=3|    |LS=3| |LF=9|    |LS=9| |LF=14|
+-----------+    +-----------+    +------------+
|  DRAG=n/a |    |  DRAG=n/a |    |  DRAG=n/a  |
+-----------+    +-----------+    +------------+
"""
    assert '\n'.join(aon_dia) + '\n' == expected_diagram

    aon_dia_text = p.aon_diagram_text(p.activity_on_node_diagram_data())
    assert aon_dia_text == expected_diagram
    assert p.aon_diagram_text_dump() == expected_diagram


def test_cpa_of_loaded_micro_project():
    p = cpa.Node('micro')
    p.load_network(str(CPA_MICRO_FIXTURE_PATH))

    aons = {}
    for node in sorted(p.nodes, key=operator.attrgetter('name')):
        aons[node.name] = node.activity_on_node_text()

    cp = p.get_critical_path()
    assert cp == [p.name_to_node['A'], p.name_to_node['D'], p.name_to_node['E']]
    assert p is not None
    aon_p = """\
+----------------+
|     DUR=14     |
+----------------+
|ES=0|     |EF=14|
|----|micro|-----|
|LS=0|     |LF=14|
+----------------+
|    DRAG=n/a    |
+----------------+
"""
    assert p.activity_on_node_text() == aon_p
    p_aon_map = {
        'duration': 14,
        'earliest_start': 0,
        'earliest_finish': 14,
        'name': 'micro',
        'latest_start': 0,
        'latest_finish': 14,
        'drag': None,
    }
    the_data = p.activity_on_node()
    assert the_data == p_aon_map
    assert json.dumps(the_data) == json.dumps(p_aon_map)  # Just to be sure :-)

    assert p.activity_on_node_strings() == tuple(row for row in aon_p.split('\n') if row)

    expected_diagram = """\
+-----------+    +-----------+    +------------+
|   DUR=3   |    |   DUR=6   |    |   DUR=5    |
+-----------+    +-----------+    +------------+
|ES=0| |EF=3|    |ES=3| |EF=9|    |ES=9| |EF=14|
|----|A|----| => |----|D|----| => |----|E|-----|
|LS=0| |LF=3|    |LS=3| |LF=9|    |LS=9| |LF=14|
+-----------+    +-----------+    +------------+
|  DRAG=n/a |    |  DRAG=n/a |    |  DRAG=n/a  |
+-----------+    +-----------+    +------------+
"""
    aon_dia_text = p.aon_diagram_text(p.activity_on_node_diagram_data())
    assert aon_dia_text == expected_diagram
    assert p.aon_diagram_text_dump() == expected_diagram


def test_model_loaded_small():
    p = cpa.Node('small-project')
    p.load_network(str(CPA_SMALL_FIXTURE_PATH))

    expected_names = CPA_SMALL_CRTICAL_PATH_EXPECTED
    assert [str(member) for member in p.get_critical_path()] == expected_names  # type: ignore
    activity_on_project = """\
+--------------------------+
|         DUR=8346         |
+--------------------------+
|ES=0|             |EF=8346|
|----|small-project|-------|
|LS=0|             |LF=8346|
+--------------------------+
|         DRAG=n/a         |
+--------------------------+
"""
    assert p.activity_on_node_text() == activity_on_project

    activity_on_node_4696 = """\
+--------------------+
|       DUR=31       |
+--------------------+
|ES=8305|    |EF=8336|
|-------|4696|-------|
|LS=8305|    |LF=8336|
+--------------------+
|      DRAG=n/a      |
+--------------------+
"""
    assert p.lookup_node('4696').activity_on_node_text() == activity_on_node_4696

    activity_on_node_4706 = """\
+--------------------+
|       DUR=10       |
+--------------------+
|ES=8336|    |EF=8346|
|-------|4706|-------|
|LS=8336|    |LF=8346|
+--------------------+
|      DRAG=n/a      |
+--------------------+
"""
    assert p.lookup_node('4706').activity_on_node_text() == activity_on_node_4706

    activity_on_node_4707 = """\
+--------------------+
|       DUR=0        |
+--------------------+
|ES=8346|    |EF=8346|
|-------|4707|-------|
|LS=8346|    |LF=8346|
+--------------------+
|      DRAG=n/a      |
+--------------------+
"""
    assert p.lookup_node('4707').activity_on_node_text() == activity_on_node_4707


prosessilouhinta/cpa.py
"""Naive Critical Path Analysis (CPA) implementation."""
import json
import pathlib
from typing import no_type_check  # Self when 3.11 is lowest bound

ActOnNodeMap = dict[str, str | int | float | None]
ActOnNodeTup = tuple[str, str, str, str, str, str, str, str, str]
ENCODING = 'utf-8'


@no_type_check
def cyclic(digraph) -> bool:
    """Determine if the directed graph has a cycle per a stack of iterators.

    The dgraph structure maps vertices to iterables of neighbouring vertices.

    # Examples:

    >>> cyclic({})
    False
    >>> cyclic({'a': ('a',)})
    True
    >>> cyclic({'a': ('z',), 'z': ('b',), 'b': ('y', 'z')})
    True
    >>> cyclic({'a': ('z',), 'z': ('b',), 'b': ('y', 'c')})
    False
    """
    visited = set()
    path = []
    paths = set(path)
    sit = [iter(digraph)]
    while sit:
        for vertex in sit[-1]:
            if vertex in paths:
                return True
            elif vertex not in visited:
                visited.add(vertex)
                path.append(vertex)
                paths.add(vertex)
                sit.append(iter(digraph.get(vertex, tuple())))
                break
        else:
            paths.discard(path.pop() if path else None)
            sit.pop()
    return False


@no_type_check
class Node:
    """Represents a task with linked nodes in an act precedence network."""

    @no_type_check
    def __init__(self, name, duration=None, lag=0) -> None:
        """Initialize the model."""
        self.parent: Node | None = None
        self.name: str = name  # model.NAME shall be unique
        self.description: str | None = None
        self.duration: float | None = duration  # model.DUR in agreed time units
        self.lag: int | float | None = lag  # preceding.finished plus in agreed time units
        self.drag: int | float | None = None  # model.DRAG in agreed time units
        self._es: int | float | None = None  # model.ES
        self.ef: int | float | None = None  # model.EF
        self.ls: int | float | None = None  # model.LS
        self.lf: int | float | None = None  # model.LF
        self._free_float: int | float | None = None  # act can delay without change start of any other act
        self._total_float: int | float | None = None  # act can delay without increase of overall project dur
        self.nodes: list[Node] = []
        self.name_to_node: dict[str, Node] = {}
        self.to_nodes: set[Node] = set()
        self.incoming_nodes: set[Node] = set()
        self.forward_pending: set[Node] = set()
        self.backward_pending: list[Node] = []
        self._critical_path = None
        self.exit_node: Node | None = None

    @no_type_check
    def lookup_node(self, name: str):
        return self.name_to_node[name]

    @no_type_check
    def get_or_create_node(self, name, **kwargs):
        try:
            return self.lookup_node(name=name)
        except KeyError:
            n = Node(name=name, **kwargs)
            self.add(n)
            return n

    @property
    def es(self) -> int | float | None:
        return self._es

    @es.setter
    def es(self, v: int | float) -> None:
        self._es = v
        if self.parent:
            self.parent.forward_pending.add(self)

    def __repr__(self) -> str:
        return str(self.name)

    def __hash__(self) -> int:
        return hash(self.name)

    @no_type_check
    def __eq__(self, other) -> bool:
        if not isinstance(other, Node):
            return False
        return self.name == other.name

    @no_type_check
    def __ne__(self, other) -> bool:
        return not self.__eq__(other)

    @no_type_check
    def add(self, node):
        """Includes the given node as a child node."""
        if not isinstance(node, Node):
            raise ValueError(f'tried to add non-Node instance {type(node).__name__}')
        if node.duration is None:
            raise ValueError('unspecified duration')
        if node in self.nodes:
            return
        self.nodes.append(node)
        self.name_to_node[node.name] = node
        node.parent = self
        self.forward_pending.add(node)
        self._critical_path = None
        return node

    @no_type_check
    def link(self, from_node, to_node=None) -> 'Node':
        """Directed link of two child nodes in the graph."""
        if not isinstance(from_node, Node):
            from_node = self.name_to_node[from_node]
        if not isinstance(from_node, Node):
            raise ValueError(f'tried to link from non-Node instance {type(from_node).__name__}')
        if to_node is not None:
            if not isinstance(to_node, Node):
                to_node = self.name_to_node[to_node]
            if not isinstance(to_node, Node):
                raise ValueError(f'tried to link to non-Node instance {type(to_node).__name__}')
            from_node.to_nodes.add(to_node)
            to_node.incoming_nodes.add(from_node)
        else:
            self.to_nodes.add(from_node)
            from_node.incoming_nodes.add(self)
        return self

    @no_type_check
    def load_network(self, file_path: str) -> None:
        """Load the network from JSON file with matching top level name value."""
        if not file_path.strip():
            raise ValueError('cannot load from empty file path string')
        fp = pathlib.Path(file_path)
        if not fp.is_file():
            raise ValueError('cannot load from non existing file')
        if not fp.stat().st_size:
            raise ValueError('cannot load from empty file')
        with open(fp, 'rt', encoding=ENCODING) as handle:
            network = json.load(handle)
        if not network:
            raise ValueError('cannot load from empty network')
        name = network.get('name', 'not-existing-name-key')
        if name != self.name:
            raise ValueError(f'cannot load network with name ({name}) into ({self.name})')
        x_nodes = {}
        for x_key, x_node in network.get('nodes', {}).items():
            duration = x_node.get('duration', None)
            lag = x_node.get('lag', 0)
            x_nodes[x_key] = self.add(Node(x_key, duration=duration, lag=lag))
        for edge in network.get('edges', []):
            if len(edge) < 2:
                if not edge:
                    raise ValueError('cannot build an empty edge')
                raise ValueError(f'cannot build directed edge from ({edge[0]}) without a target')
            if len(edge) == 2:
                if any(edg not in x_nodes for edg in edge):
                    raise ValueError(f'cannot build edge with nodes not in {x_nodes}')
                self.link(x_nodes[edge[0]], x_nodes[edge[1]])
            else:
                raise NotImplementedError('Compressed edges notation not yet implemented.')
        self.update_all()

    @property
    @no_type_check
    def first_nodes(self):
        """Child nodes that have no in-bound dependencies."""
        first = set(self.nodes)
        for node in self.nodes:
            first.difference_update(node.to_nodes)
        return first

    @property
    @no_type_check
    def last_nodes(self):
        """Child nodes that have to out-bound dependencies."""
        return [node for node in self.nodes if not node.to_nodes]

    @no_type_check
    def update_forward(self) -> None:
        """Updates forward timing calculations for the current node assuming min(es) is set."""
        changed = False
        if self.es is not None and self.duration is not None:
            self.ef = self.es + self.duration
            changed = True

        if changed:
            for to_node in self.to_nodes:
                if to_node == self:
                    continue
                add_some_lag = self.ef + to_node.lag
                to_node.es = add_some_lag if to_node.es is None else max(to_node.es, add_some_lag)

                if self.parent:
                    self.parent.forward_pending.add(to_node)

            if self.parent:
                self.parent.backward_pending.append(self)

    @no_type_check
    def update_backward(self) -> None:
        """Updates backward timing calculations for the current node."""
        if self.lf is None:
            if self.to_nodes:
                self.lf = min(target.ls for target in self.to_nodes)
            else:
                self.lf = self.ef
            if self.lf is None:
                raise ValueError('No latest finish time found')
        self.ls = self.lf - self.duration

    @no_type_check
    def add_exit(self) -> None:
        """Links all leaf nodes to a common exit node."""
        if self.exit_node is None:
            self.exit_node = Node('COMMON_EXIT', duration=0)
            self.add(self.exit_node)
        for node in self.nodes:
            if node is self.exit_node:
                continue
            if not node.to_nodes:
                self.link(from_node=node, to_node=self.exit_node)

    @no_type_check
    def update_all(self) -> None:
        """Updates timing calculations for all children nodes."""
        if not self.is_acyclic():
            raise TypeError('Network contains cycles')

        for node in list(self.forward_pending.intersection(self.first_nodes)):
            node.es = self.lag + node.lag
            node.update_forward()
            self.forward_pending.remove(node)

        forward_priors = set()
        while self.forward_pending:
            q = set(self.forward_pending)
            self.forward_pending.clear()
            while q:
                node = q.pop()
                if node in forward_priors:
                    continue
                node.update_forward()

        backward_priors = set()
        while self.backward_pending:
            node = self.backward_pending.pop()
            if node in backward_priors:
                continue
            node.update_backward()

        duration, path, priors = self.get_critical_path(as_item=True)
        self._critical_path = duration, path, priors
        self.duration = duration
        self.es = path[0].es
        self.ls = path[0].ls
        self.ef = path[-1].ef
        self.lf = path[-1].lf

    @no_type_check
    def get_critical_path(self, as_item: bool = False):
        """Finds the longest path in among the child nodes."""
        if self._critical_path is not None:
            return self._critical_path[1]
        longest = None
        q = [(_.duration, [_], set([_])) for _ in self.first_nodes]
        while q:
            item = length, path, priors = q.pop(0)
            if longest is None:
                longest = item
            else:
                try:
                    longest = max(longest, item)
                except TypeError:
                    longest = longest
            for to_node in path[-1].to_nodes:
                if to_node in priors:
                    continue
                q.append((length + to_node.duration, path + [to_node], priors.union([to_node])))
        if longest is None:
            return
        elif as_item:
            return longest
        else:
            return longest[1]

    @no_type_check
    def activity(self, node) -> ActOnNodeMap:
        """Provide the activity-on-node element of given node as JSON serializable dict."""
        return {
            'duration': node.duration,
            'earliest_start': node.es,
            'earliest_finish': node.ef,
            'name': node.name,
            'latest_start': node.ls,
            'latest_finish': node.lf,
            'drag': node.drag,
        }

    @no_type_check
    def activity_on_node(self) -> ActOnNodeMap:
        """Provide the activity-on-node element as JSON serializable dict."""
        return self.activity(self)

    @no_type_check
    def activity_on_node_diagram_data(self) -> list[ActOnNodeMap]:
        """Provide the activity-on-node diagram as JSON serializable list of dicts."""
        cp = self.get_critical_path()
        return [self.activity(node) for node in cp] if cp else []

    @staticmethod
    @no_type_check
    def aon_strings(aon: ActOnNodeMap) -> ActOnNodeTup:
        """Provide the activity-on-node element from map data as 9-tuple of strings.

        Example:

        cf. example of activity_on_node_strings method
        """
        duration = aon['duration']
        earliest_start = aon['earliest_start']
        earliest_finish = aon['earliest_finish']
        name = aon['name']
        latest_start = aon['latest_start']
        latest_finish = aon['latest_finish']
        drag = aon['drag']

        lk_width = max(len(f'{lk}=') for lk in ('ES', 'EF', 'LS', 'LF'))
        sp, hr, vr = ' ', '-', '|'
        bd_width = len(vr)
        max_width, lc_max_width, cc_max_width, rc_max_width = 0, 0, 0, 0
        lc_max_width = max(len(str(earliest_start)), len(str(latest_start))) + lk_width
        cc_max_width = len(str(name))
        rc_max_width = max(len(str(earliest_finish)), len(str(latest_finish))) + lk_width
        max_width = sum((lc_max_width, cc_max_width, rc_max_width)) + 2 * bd_width  # inner borders

        section_sep = f'+{hr * max_width}+'

        dur_disp = f'DUR={duration}'
        dur_sect = f'{vr}{dur_disp.center(max_width, sp)}{vr}'
        es_disp = f'ES={earliest_start}'
        ef_disp = f'EF={earliest_finish}'
        es_ef_sects = (
            f'{vr}{es_disp.center(lc_max_width, sp)}'
            f'{vr}{sp.center(cc_max_width, sp)}'
            f'{vr}{ef_disp.center(rc_max_width, sp)}{vr}'
        )
        name_sect = (
            f'{vr}{hr.center(lc_max_width, hr)}'
            f'{vr}{str(name).center(cc_max_width, sp)}'
            f'{vr}{hr.center(rc_max_width, hr)}{vr}'
        )
        ls_disp = f'LS={latest_start}'
        lf_disp = f'LF={latest_finish}'
        ls_lf_sects = (
            f'{vr}{ls_disp.center(lc_max_width, sp)}'
            f'{vr}{sp.center(cc_max_width, sp)}'
            f'{vr}{lf_disp.center(rc_max_width, sp)}{vr}'
        )
        drg_disp = f'DRAG={drag if drag is not None else "n/a"}'
        drg_sect = f'{vr}{drg_disp.center(max_width, sp)}{vr}'

        return (
            section_sep,
            dur_sect,
            section_sep,
            es_ef_sects,
            name_sect,
            ls_lf_sects,
            section_sep,
            drg_sect,
            section_sep,
        )

    def activity_on_node_strings(self) -> ActOnNodeTup:
        """Provide the activity-on-node element as 9-tuple of strings for e.g. linking along the critical path.

        Example:

        +-----------------------+  # section_sep
        |        DUR=31         |  # dur_sect
        +-----------------------+  # section_sep
        |ES=8308|       |EF=8339|  # es_ef_sects
        |-------|  4696 |-------|  # name_sect
        |LS=8308|       |LF=8339|  # ls_lf_sects
        +-----------------------+  # section_sep
        |       DRAG=None       |  # drg_sect
        +-----------------------+  # section_sep
        """
        lk_width = max(len(f'{lk}=') for lk in ('ES', 'EF', 'LS', 'LF'))
        sp, hr, vr = ' ', '-', '|'
        bd_width = len(vr)
        max_width, lc_max_width, cc_max_width, rc_max_width = 0, 0, 0, 0
        lc_max_width = max(len(str(self.es)), len(str(self.ls))) + lk_width
        cc_max_width = len(str(self.name))
        rc_max_width = max(len(str(self.ef)), len(str(self.lf))) + lk_width
        max_width = sum((lc_max_width, cc_max_width, rc_max_width)) + 2 * bd_width  # inner borders

        section_sep = f'+{hr * max_width}+'

        dur_disp = f'DUR={self.duration}'
        dur_sect = f'{vr}{dur_disp.center(max_width, sp)}{vr}'
        es_disp = f'ES={self.es}'
        ef_disp = f'EF={self.ef}'
        es_ef_sects = (
            f'{vr}{es_disp.center(lc_max_width, sp)}'
            f'{vr}{sp.center(cc_max_width, sp)}'
            f'{vr}{ef_disp.center(rc_max_width, sp)}{vr}'
        )
        name_sect = (
            f'{vr}{hr.center(lc_max_width, hr)}'
            f'{vr}{str(self.name).center(cc_max_width, sp)}'
            f'{vr}{hr.center(rc_max_width, hr)}{vr}'
        )
        ls_disp = f'LS={self.ls}'
        lf_disp = f'LF={self.lf}'
        ls_lf_sects = (
            f'{vr}{ls_disp.center(lc_max_width, sp)}'
            f'{vr}{sp.center(cc_max_width, sp)}'
            f'{vr}{lf_disp.center(rc_max_width, sp)}{vr}'
        )
        drg_disp = f'DRAG={self.drag if self.drag is not None else "n/a"}'
        drg_sect = f'{vr}{drg_disp.center(max_width, sp)}{vr}'

        return (
            section_sep,
            dur_sect,
            section_sep,
            es_ef_sects,
            name_sect,
            ls_lf_sects,
            section_sep,
            drg_sect,
            section_sep,
        )

    @staticmethod
    def aon_strings_with_link(aon_strings: ActOnNodeTup) -> ActOnNodeTup:
        """DRY."""
        arrow = ' => '
        spacer = ' ' * len(arrow)
        link: ActOnNodeTup = (spacer, spacer, spacer, spacer, arrow, spacer, spacer, spacer, spacer)
        return tuple(f'{rec}{spc}' for rec, spc in zip(aon_strings, link))  # type: ignore

    @staticmethod
    def aon_diagram_text(aons: list[ActOnNodeMap]) -> str:
        """Build a text representation for a path with activity-on-nodes linked directionally."""
        dia = []
        for aon in aons[:-1]:
            dia.append(Node.aon_strings_with_link(Node.aon_strings(aon)))
        dia.append(Node.aon_strings(aons[-1]))
        return '\n'.join(''.join(row) for row in zip(*dia)) + '\n'

    def aon_diagram_text_dump(self) -> str:
        """Build a text representation for the critical path with activity-on-nodes linked directionally."""
        return self.aon_diagram_text(self.activity_on_node_diagram_data())

    def activity_on_node_strings_with_link(self) -> ActOnNodeTup:
        """DRY."""
        return self.aon_strings_with_link(self.activity_on_node_strings())

    def activity_on_node_text(self) -> str:
        """Provide the activity-on-node element as string with trailing newline for e.g. a textual diagram."""
        return '\n'.join(self.activity_on_node_strings()) + '\n'

    def is_acyclic(self) -> bool:
        g = dict((node.name, tuple(child.name for child in node.to_nodes)) for node in self.nodes)
        return not cyclic(g)
